presto-main/src/main/java/com/facebook/presto/operator/BlockEncodingBuffers.java

```diff
@@ -0,0 +1,442 @@
+        }
+
+        // Process the next BITS_IN_BYTE * n positions. We now have processed (offset - positionsOffset) positions
+        int remainingPositions = batchSize - (offset - positionsOffset);
+        int positionsToEncode = remainingPositions & ~0b111;
+        for (int i = 0; i < positionsToEncode; i += BITS_IN_BYTE) {
+            byte value = 0;
+            value |= block.isNull(positions[offset]) ? 0b1000_0000 : 0;
+            value |= block.isNull(positions[offset + 1]) ? 0b0100_0000 : 0;
+            value |= block.isNull(positions[offset + 2]) ? 0b0010_0000 : 0;
+            value |= block.isNull(positions[offset + 3]) ? 0b0001_0000 : 0;
+            value |= block.isNull(positions[offset + 4]) ? 0b0000_1000 : 0;
+            value |= block.isNull(positions[offset + 5]) ? 0b0000_0100 : 0;
+            value |= block.isNull(positions[offset + 6]) ? 0b0000_0010 : 0;
+            value |= block.isNull(positions[offset + 7]) ? 0b0000_0001 : 0;
+
+            nullsBufferContainsNull |= (value != 0);
+            nullsBufferIndex = setByte(nullsBuffer, nullsBufferIndex, value);
+            offset += BITS_IN_BYTE;
+        }
+
+        // Process the remaining positions
+        remainingPositions &= 0b111;
+        while (remainingNullsCount < remainingPositions) {
+            remainingNullsFromLastBatch[remainingNullsCount++] = block.isNull(positions[offset++]);
+        }
+    }
+
+    private void encodeNonNullsAsBits(int count)
+    {
+        if (remainingNullsCount + count < BITS_IN_BYTE) {
+            // just put all of this batch to remainingNullsFromLastBatch
+            for (int i = 0; i < count; i++) {
+                remainingNullsFromLastBatch[remainingNullsCount++] = false;
+            }
+            return;
+        }
+
+        int remainingPositions = count - encodeRemainingNullsAsBits();
+        int bytesCount = remainingPositions >>> 3;
+
+        nullsBufferIndex = fill(nullsBuffer, nullsBufferIndex, bytesCount, (byte) 0);
+
+        remainingPositions &= 0b111;
+        while (remainingNullsCount < remainingPositions) {
+            remainingNullsFromLastBatch[remainingNullsCount++] = false;
+        }
+    }
+
+    private int encodeRemainingNullsAsBits()
+    {
+        if (remainingNullsCount == 0) {
+            return 0;
+        }
+
+        byte value = 0;
+        for (int i = 0; i < remainingNullsCount; i++) {
+            value |= remainingNullsFromLastBatch[i] ? 0b1000_0000 >>> i : 0;
+        }
+
+        nullsBufferContainsNull |= (value != 0);
+        nullsBufferIndex = setByte(nullsBuffer, nullsBufferIndex, value);
+
+        int padding = BITS_IN_BYTE - remainingNullsCount;
+        remainingNullsCount = 0;
+        return padding;
+    }
+
+    protected long getNullsBufferSizeInBytes()
+    {
+        return nullsBufferIndex +                               // nulls buffer
+                remainingNullsCount;   // the remaining nulls not serialized yet
```

Comment:
Is this correct? All of the remaining nulls will be encoded into one byte, hence, I'd expect this to be `+1`.

Reply:
I actually wasn't quite sure if we need to use remainingNullsCount or just 8, but I don't think it should be 1. In my understanding sizeInBytes usually represents the logical size of the objects. Here it includes the nullsBuffer and remainingNulls. remainingNulls is array of boolean of size 8, but it may contain < 8 values. And in our case we use sizeInBytes to simulate retainedSizeInBytes because the latter is more expensive, so I think it should be at least remainingNullsCount.

Reply:
@yingsu00 Thanks for clarifying. Looks like the only purpose of this method is to approximate the retained size in bytes. If that's the case, I suggest to rename this (and similar) methods to `getApproximateRetainedSizeInBytes`. This would make it very clear that this method should return about the same result as `getRetainedSizeInBytes`, but try to burn less CPU.

Also, I'm seeing that `PagePartitioner#getRetainedSizeInBytes` is called only once to initialize `partitionsInitialRetainedSize`. At this point `decodedBlocks` array contains all nulls, hence, the following loop is not needed:

```
            for (int i = 0; i < decodedBlocks.length; i++) {
                retainedSizeInBytes += decodedBlocks[i] == null ? 0 : decodedBlocks[i].getRetainedSizeInBytes();
            }
```

If you remove this loop, `DecodedBlockNode#getRetainedSizeInBytes` will become unused and can be removed as well.

By the same logic, `PartitionData#getRetainedSizeInBytes` is always called with `blockEncodingBuffers` being null, hence, it can be simplified to remove calls to `BlockEncodingBuffers.getRetainedSizeInBytes`. Then these methods can be removed from BlockEncodingBuffers as well.

Finally, `OptimizedPartitionedOutputOperatorFactory.pageFunction` variable has a confusing name. I think it should be `pagePartitioner` because `PageFunction` is something else.

Reply:
The logic of using getSizeInBytes() instead of getRetainedSizeInBytes() was copied from PartitionedOutputOperator. However when I ran the benchmarks, I didn't see any difference in terms of performance of these two methods. So I removed getSizeInBytes() and kept getRetainedSizeInBytes() because it's the actual physical memory we should be tracking. Also introduced getInitialRetainedSizeInBytes() and removed the null check in getRetainedSizeInBytes(). Removed the initialRetainedSizeInBytes because we don't need this approximation since we use getRetainedSizeInBytes() now.

For the last point, did you mean partitionFunction? 
`private final PagePartitioner partitionFunction;`
It's also copied from PartitionedOutputOperator. I renamed it to pagePartitioner.